src/backend/utils/gp/segadmin.c

```diff
@@ -194,13 +194,27 @@ remove_segment_config(int16 dbid)
 				Anum_gp_segment_configuration_dbid,
 				BTEqualStrategyNumber, F_INT2EQ,
 				Int16GetDatum(dbid));
-
-	sscan = systable_beginscan(rel, GpSegmentConfigDbidIndexId, true,
+	sscan = systable_beginscan(rel, GpSegmentConfigDbidWarehouseIndexId, true,
 							   NULL, 1, &scankey);
 	while ((tuple = systable_getnext(sscan)) != NULL)
 	{
-		CatalogTupleDelete(rel, &tuple->t_self);
-		numDel++;
+		Datum		attr;
+		bool		isNull;
+		char	   *warehouse_name = NULL;
+
+		attr = heap_getattr(tuple, Anum_gp_segment_configuration_warehouse_name,
+							RelationGetDescr(rel), &isNull);
+		if (!isNull)
+			warehouse_name = TextDatumGetCString(attr);
+
+		attr = heap_getattr(tuple, Anum_gp_segment_configuration_content,
+							RelationGetDescr(rel), &isNull);
+		Assert(!isNull);
+		if (DatumGetInt16(attr) == MASTER_CONTENT_ID || strcmp(warehouse_name, current_warehouse) == 0)
```

Comment:
why need this  (DatumGetInt16(attr) == MASTER_CONTENT_ID) condition？

Reply:
warehouse_name of the master is null, when we query from the `gp_sgment_configuration`, this clause is like `where content == -1 or current_warehouse = warehouse_name`.